src/lib/storage/encoded_columns/base_dictionary_column.hpp

```diff
@@ -0,0 +1,47 @@
+#pragma once
+
+#include <limits>
+#include <memory>
+
+#include "storage/encoded_columns/base_encoded_column.hpp"
+
+namespace opossum {
+
+class BaseZeroSuppressionVector;
+
+/**
+ * @brief Base class of DictionaryColumn<T> exposing type independent interface
+ */
+class BaseDictionaryColumn : public BaseEncodedColumn {
+ public:
+  EncodingType encoding_type() const final { return EncodingType::Dictionary; }
+
+  /**
+   * @brief Returns index (i.e. ValueID) of first dictionary entry >= search value
+   *
+   * @param value the search value
+   * @return INVALID_VALUE_ID if entries are smaller than value
+   */
+  virtual ValueID lower_bound(const AllTypeVariant& value) const = 0;
+
+  /**
+   * @brief Returns index (i.e. ValueID) of first dictionary entry > search value
+   *
+   * @param value the search value
+   * @return INVALID_VALUE_ID if entries are smaller than or equal to value
+   */
+  virtual ValueID upper_bound(const AllTypeVariant& value) const = 0;
+
+  /**
+   * @brief The size of the dictionary
+   */
+  virtual size_t unique_values_count() const = 0;
+
+  virtual std::shared_ptr<const BaseZeroSuppressionVector> attribute_vector() const = 0;
+
+  /**
+   * @brief Returns encoding specific null value ID
```

Comment:
Why do we have encoding-specific null ids?

Reply:
We already had them before, but the virtual interface of `BaseAttributeVector` hid this from the user with the cost of having a branch in each `get(ValueID)`. The way we did in the old implementation is however not generalisable. SIMD-BP128 for example has a variable bit size so the maximum possible value depends on the segment. So the easiest solution was to set the null value ID to dictionary size + 1. This also avoids also the unnecessary conversion inside of `BaseAttributeVector`.